pkg/ruler/compat.go

```diff
@@ -163,6 +163,29 @@ func MetricsQueryFunc(qf rules.QueryFunc, queries, failedQueries prometheus.Coun
 	}
 }
 
+func RecordAndReportRuleQueryMetrics(qf rules.QueryFunc, queryTime prometheus.Counter, logger log.Logger) rules.QueryFunc {
+	return func(ctx context.Context, qs string, t time.Time) (promql.Vector, error) {
+		// If we've been passed a counter we want to record the wall time spent executing this request.
+		if queryTime != nil {
+			timer := prometheus.NewTimer(nil)
+			defer func() {
+				querySeconds := timer.ObserveDuration().Seconds()
+				queryTime.Add(querySeconds)
+
+				// Log ruler query stats.
+				logMessage := append([]interface{}{
+					"msg", "ruler query stats",
+					"cortex_ruler_query_seconds_total", querySeconds,
+				}, qs)
+				level.Info(util_log.WithContext(ctx, logger)).Log(logMessage...)
+			}()
+		}
+
+		result, err := qf(ctx, qs, t)
+		return result, err
+	}
+}
```

Comment:
We can simplify this function:
1) we don't need to return new function, if `queryTime` is nil
2) not sure what's going on with logging here, but we can just call `Log`, right?

```suggestion
func RecordAndReportRuleQueryMetrics(qf rules.QueryFunc, queryTime prometheus.Counter, logger log.Logger) rules.QueryFunc {
	if queryTime == nil {
		return qf
	}

	return func(ctx context.Context, qs string, t time.Time) (promql.Vector, error) {
		timer := prometheus.NewTimer(nil)
		defer func() {
			querySeconds := timer.ObserveDuration().Seconds()
			queryTime.Add(querySeconds)

			level.Info(util_log.WithContext(ctx, logger)).Log("msg", "ruler query stats", "cortex_ruler_query_seconds_total", querySeconds, "query", qs)
		}()

		result, err := qf(ctx, qs, t)
		return result, err
	}
}
```

Reply:
>not sure what's going on with logging here, but we can just call Log, right?

I'm trying to follow what we do for [query front-end logging ](https://github.com/cortexproject/cortex/blob/95fedaa11c21e0ba2090ab4c6bc9431017fe11f3/pkg/frontend/transport/handler.go#L174-L182) in case a user would also want to record this log message in the ruler and use it for billing instead of the added metric.